feat: month9 step1 Routing & Auth

dmitry-razumov opened the pull request:
month9 step1 Routing & Auth

dmitry-razumov committed:
feat: month9 step1 Routing & Auth

gennady-bars commented on src/utils/burger-api.js:
```diff
@@ -1,23 +1,22 @@
 
 export const getIngredientsData = async () => {
   const res = await fetch(`${BASE_URL}ingredients`);
   return await checkResponse(res);
```
**Можно лучше**

Можно сделать **универсальную** функцию запроса с **проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  function request(url, options) {
    // принимает два аргумента: урл и объект опций, как и `fetch`
    return fetch(url, options).then(checkResponse)
  }

```

И теперь просто нужно заменить все `fetch` на `request` и убрать дублирование проверки на `ok`. Все остальное будет без изменений. Код станет чище

И даже можно поместить внутрь `baseUrl`, чтобы не дублировать его в каждом запросе. Тогда нужно будет передавать просто `endpoint` в вызов

gennady-bars commented on src/services/actions/user.js:
```diff
@@ -0,0 +1,82 @@
+      localStorage.setItem("refreshToken", res.refreshToken);
+      dispatch(setUser(res.user));
+      dispatch(setAuthChecked(true));
+    });
```
Для любых запросов на сервер в **самом конце цепочки** запроса в `thunk` нужно осуществлять проверку на возможные ошибки с помощью `catch`.

**Нужно исправить это везде по коду в конце запросов.**

gennady-bars commented on src/pages/login/login.jsx:
```diff
@@ -0,0 +1,44 @@
+
+function Login() {
+  const dispatch = useDispatch();
+  const [form, setValue] = useState({email:'', password:''});
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues={}) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

gennady-bars commented on src/pages/login/login.jsx:
```diff
@@ -0,0 +1,44 @@
+	return (
+		<div className={styles.login}>
+      <h2 className={styles.title}>Вход</h2>
+      <form className={styles.form} onSubmit={onSubmit}>
```
Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. Обработка события `submit` еще срабатывает при нажатии на `Enter`, так как это встроенная логика сабмитов в браузере.

gennady-bars commented on src/components/protected-route.jsx:
```diff
@@ -0,0 +1,28 @@
+    return null;
+  }
+
+  if (onlyUnAuth && user) {
+    const { from } = location.state || { from: { pathname: "/" } };
+    return <Navigate to={from} />;
+  }
+
+  if (!onlyUnAuth && !user) {
+    return <Navigate to="/login" state={{ from: location }} />;
+  }
```
Хорошо сделан `Protected`

gennady-bars commented on src/components/app/app.jsx:
```diff
@@ -1,27 +1,74 @@
   const dispatch = useDispatch();
+  const location = useLocation();
+  const navigate = useNavigate();
+  const background = location.state && location.state.background;
```
Отлично сделана логика проверки, был ли пользователь на сайте, и показывать ли страницу  с деталями, а не модально окно

gennady-bars commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -61,65 +69,61 @@ function BurgerConstructor() {
+        <div className={styles.scrollbarList}>
           <ul className={styles.items}>
              {ingredients.map((item, index) => {
                 return <BurgerItem key={index} item={item} index={index} moveItem={moveItem}/>
```
`key={index}`

Для атрибута `key` в конструкторе нужно использовать для ингредиента уникальный `uuid` (Вам нужно его генерировать во время **сброса в зону** `DnD`). Это можно делать с помощью библиотеки `uuid` https://www.npmjs.com/package/uuid , или Вам должны были показать, как это сделать, иначе некорректно может работать DND. А в `key` будет уже просто **key={ item.uniqueId }**

gennady-bars replied:
Нужно генерировать uuid при создании экшена (желательно, в **`action creator`**). И созданный uuid должен приходить в экшене в `payload` или как отдельное поле, или сразу как часть объекта ингредиента. Посоветуйтесь с наставником, пожалуйста, если не получится
```js

// вот это - `action creator`
export const addIngridient = (item) => {
    return {
        type: ADD_INGRIDIENT,
        payload: {
            ...item, // используем `spread`, чтобы поменять ссылку на объект. Таким образом `redux` обновит его в хранилище
           uniqueId: uuid4()  // и добавляем в объект новое поле, которое потом будет использовано в `key`
        }
    }
}


```

gennady-bars replied:
[Почему индекс не нужно использовать, можно прочитать тут]( https://habr.com/ru/companies/hh/articles/352150/)

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -10,21 +11,23 @@ const products = [
       <div className={styles.scrollbar}>
         { products.map((product, index) => 
-          (<ShowItems key={index} name={product.name} type={product.type} />))
+          (<ShowItems key={index} name={product.name} type={product.type} {...{setVisibleSection, containerRef}}/>))
```
` key={index} `


Для атрибута `key` лучше всего использовать уникальный `id` : у карточки как раз есть такой только с нижним подчеркиванием `_id`
[Почему индекс не нужно использовать, можно прочитать тут]( https://habr.com/ru/companies/hh/articles/352150/)

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку.
* Хорошо сделан `Protected`
* Отлично сделана логика проверки, был ли пользователь на сайте, и показывать ли страницу  с деталями, а не модально окно
* Хороший роутинг


но есть некоторые недочеты:



* Для любых запросов на сервер в **самом конце цепочки** запроса в `thunk` нужно осуществлять проверку на возможные ошибки с помощью `catch`.
* Для атрибута `key` в конструкторе нужно использовать для ингредиента уникальный `uuid` (Вам нужно его генерировать во время **сброса в зону** `DnD`). Это можно делать с помощью библиотеки `uuid` https://www.npmjs.com/package/uuid , или Вам должны были показать, как это сделать, иначе некорректно может работать DND. А в `key` будет уже просто **key={ item.uniqueId }**




**Можно лучше**


* Желательно **очищать конструктор** после успешного получения номера заказа с сервера в блоке `then`, чтобы пользователь мог следующий заказ сделать, не удаляя старые ингредиенты

dmitry-razumov committed:
refact: fetch's обернуты в request + сообщения http ошибок (4xx) прокинуты в store и формы
fix: added catch в thunk'и
feat: useForm hook
fix: key={uuid} в конструкторе
fix: уникальные id в key в ингридиентах
feat: clear конструктор бургера при успешном заказе

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -18,16 +18,16 @@ function BurgerIngredients() {
-        { products.map((product, index) => (
-          <Tab value={product.name} key={index} active={visibleSection === product.type} >
+        { products.map((product) => (
+          <Tab value={product.name} key={uuid4()} active={visibleSection === product.type} >
```
`key={uuid4()}` нельзя динамически создавать что-то в `key`. Для атрибута `key` нужно использовать только что-то постоянное, что всегда есть при элементе, иначе реакт не поймет, что этот компонент уже был отрисован и ничего в нем не изменилось (реакт начнет подгружать данные, картинки и тд из-за этого). Для атрибута `key` обычно используют уникальный `_id`. Поищите такое же в этих данных (в крайнем случае можно использовать `index` массива, если совсем ничего уникального  нет, но индекс нельзя использовать для `key`, если есть логика `DnD`)

dmitry-razumov replied:
про динамический uuid спасибо, как-то не подумал. вернул индекс
products - это контейнеры категорий ингредиентов (булки, соусы, начинки)
индекс products не используется для DnD, просто мне показалось красивее с точки зрения кода их через мап отрендерить, чем расписывать вручную каждую категорию одинаковой версткой
```
const products = [
  {type: 'bun', name: 'Булки'},
  {type: 'sauce', name: 'Соусы'},
  {type: 'main', name : 'Начинки'}
]
```
сами отдельные ингридиенты (IngredientItem) внутри компонента ShowItems рендерятся, каждый со своим уникальным _id, получаемым с компонентом с сервера
и именно уже он используется в DnD

gennady-bars replied:
Можно   `name` или `type` использовать  в `key`

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -18,16 +18,16 @@ function BurgerIngredients() {
-        { products.map((product, index) => 
-          (<ShowItems key={index} name={product.name} type={product.type} {...{setVisibleSection, containerRef}}/>))
+        { products.map((product) => 
+          (<ShowItems key={uuid4()} name={product.name} type={product.type} {...{setVisibleSection, containerRef}}/>))
```
`key={uuid4()}` нельзя динамически создавать что-то в `key`. Для атрибута `key` нужно использовать только что-то постоянное, что всегда есть при элементе, иначе реакт не поймет, что этот компонент уже был отрисован и ничего в нем не изменилось (реакт начнет подгружать данные, картинки и тд из-за этого). Для атрибута `key` обычно используют уникальный `_id`. Поищите такое же в этих данных (в крайнем случае можно использовать `index` массива, если совсем ничего уникального  нет, но индекс нельзя использовать для `key`, если есть логика `DnD`)

dmitry-razumov replied:
вернул индекс

gennady-bars replied:
А у табов есть имя уникальное. Поэтому в `key` можно его использовать

gennady-bars reviewed (commented):
Уже почти все отлично, но:


* `key={uuid4()}` нельзя динамически создавать что-то в `key`. Для атрибута `key` нужно использовать только что-то постоянное, что всегда есть при элементе, иначе реакт не поймет, что этот компонент уже был отрисован и ничего в нем не изменилось (реакт начнет подгружать данные, картинки и тд из-за этого). Для атрибута `key` обычно используют уникальный `_id`. Поищите такое же в этих данных (в крайнем случае можно использовать `index` массива, если совсем ничего уникального  нет, но индекс нельзя использовать для `key`, если есть логика `DnD`)
* Перестала работать логика переключения табов при скролле ингредиентов. Идет какой-то конфликт

dmitry-razumov committed:
fix: вернул index для tab'ов и категорий ингредиентов (show items)

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились. С наступающим)

Удачного дальнейшего обучения.

dmitry-razumov committed:
fix: key's tabs & showItems
